Add `es2023` environment

### ESLint version

8.43.0

### What problem do you want to solve?

Node.JS 18 which supports ECMAScript 2023 has been out since April 2022. However, right now [the docs](https://eslint.org/docs/latest/use/configure/language-options) only goes up to es2022 and I think ESLint is the one showing this error:

```
error TS6046: Argument for '--target' option must be: 'es3', 'es5', 'es6', 'es2015', 'es2016', 'es2017', 'es2018', 'es2019', 'es2020', 'es2021', 'es2022', 'esnext'.
```

### What do you think is the correct solution?

Add support for es2023 :)

### Participation

- [ ] I am willing to submit a pull request for this change.

### Additional comments

_No response_

Hi @cristan, thanks for the issue! 

> Node.JS 18 which supports ECMAScript 2023 has been out since April 2022. However, right now [the docs](https://eslint.org/docs/latest/use/configure/language-options) only goes up to es2022

ESLint fully supports ECMAScript 2023. We just didn't add `es2023` environment as there are no new globals in es2023 compared to es2022. If you enable `es2022` environment and set `parserOptions.ecmaVersion` to `2023`, all ECMAScript 2023 syntax and globals will be available:

```json
{
    "env": {
        "es2022": true
    },
    "parserOptions": {
        "ecmaVersion": 2023
    }
}

```

I think we should add `es2023` environment for consistency and to avoid confusion.

> error TS6046: Argument for '--target' option must be: 'es3', 'es5', 'es6', 'es2015', 'es2016', 'es2017', 'es2018', 'es2019', 'es2020', 'es2021', 'es2022', 'esnext'.

This error is coming from TypeScript, not ESLint.

--- a/docs/src/use/configure/language-options.md
+++ b/docs/src/use/configure/language-options.md
@@ -26,6 +26,8 @@ An environment provides predefined global variables. The available environments 
 * `es2020` - adds all ECMAScript 2020 globals and automatically sets the `ecmaVersion` parser option to 11.
 * `es2021` - adds all ECMAScript 2021 globals and automatically sets the `ecmaVersion` parser option to 12.
 * `es2022` - adds all ECMAScript 2022 globals and automatically sets the `ecmaVersion` parser option to 13.
+* `es2023` - adds all ECMAScript 2023 globals and automatically sets the `ecmaVersion` parser option to 14.
+* `es2024` - adds all ECMAScript 2024 globals and automatically sets the `ecmaVersion` parser option to 15.
 * `worker` - web workers global variables.
 * `amd` - defines `require()` and `define()` as global variables as per the [amd](https://github.com/amdjs/amdjs-api/blob/master/AMD.md) spec.
 * `mocha` - adds all of the Mocha testing global variables.

--- a/package.json
+++ b/package.json
@@ -62,7 +62,7 @@
   "dependencies": {
     "@eslint-community/eslint-utils": "^4.2.0",
     "@eslint-community/regexpp": "^4.4.0",
-    "@eslint/eslintrc": "^2.0.3",
+    "@eslint/eslintrc": "github:eslint/eslintrc",
     "@eslint/js": "8.43.0",
     "@humanwhocodes/config-array": "^0.11.10",
     "@humanwhocodes/module-importer": "^1.0.1",

--- a/package.json
+++ b/package.json
@@ -62,7 +62,7 @@
   "dependencies": {
     "@eslint-community/eslint-utils": "^4.2.0",
     "@eslint-community/regexpp": "^4.4.0",
-    "@eslint/eslintrc": "github:eslint/eslintrc",
+    "@eslint/eslintrc": "^2.1.0",
     "@eslint/js": "8.43.0",
     "@humanwhocodes/config-array": "^0.11.10",
     "@humanwhocodes/module-importer": "^1.0.1",

